src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Extensions/ServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,87 @@
+using System.Linq;
+using GovUk.Education.ExploreEducationStatistics.Common.ModelBinding;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.DependencyInjection;
+using Moq;
+using static Moq.MockBehavior;
+
+namespace GovUk.Education.ExploreEducationStatistics.Common.Tests.Extensions;
+
+/// <summary>
+/// This class contains a number of extension methods for IServiceCollection that are useful when setting up
+/// integration tests that require a customised Startup class.
+/// </summary>
+public static class ServiceCollectionExtensions
+{
+    /// <summary>
+    /// This method replaces a DcContext that has been registered in Startup with an in-memory equivalent.
+    /// </summary>
+    public static IServiceCollection ReplaceDbContext<TDbContext>(this IServiceCollection services)
+        where TDbContext : DbContext
+    {
+        // Remove the default DbContext descriptor that was provided by Startup.cs.
+        var descriptor = services
+            .Single(d => d.ServiceType == typeof(DbContextOptions<TDbContext>));
+
+        services.Remove(descriptor);
+            
+        // Add the new In-Memory replacement.
+        return services.AddDbContext<TDbContext>(
+            options => options
+                .UseInMemoryDatabase(nameof(TDbContext), builder => builder.EnableNullChecks(false)));
+    }
+    
+    /// <summary>
+    /// This method replaces a service that has been registered in Startup with a new implementation. The same
+    /// lifecycle that was registered in Startup will be used to register the new service.
+    /// </summary>
+    public static IServiceCollection ReplaceService<TService>(
+        this IServiceCollection services, 
+        TService replacement)
+        where TService : class 
+    {
+        // Remove the default service descriptor that was provided by Startup.cs.
+        var descriptor = services
+            .Single(d => d.ServiceType == typeof(TService));
+    
+        services.Remove(descriptor);
+            
+        // Add the replacement.
+        return descriptor.Lifetime switch
+        {
+            ServiceLifetime.Singleton => services.AddSingleton(_ => replacement),
+            ServiceLifetime.Scoped => services.AddScoped(_ => replacement),
+            ServiceLifetime.Transient => services.AddTransient(_ => replacement),
+            _ => throw new ArgumentOutOfRangeException(
+                $"Cannot register test service with ${nameof(ServiceLifetime)} {descriptor.Lifetime}")
+        };
+    }
+    
+    /// <summary>
+    /// This method replaces a service that has been registered in Startup with a Strict Mock. The same
+    /// lifecycle that was registered in Startup will be used to register the new service.
+    /// </summary>
+    public static IServiceCollection ReplaceService<TService>(
+        this IServiceCollection services)
+        where TService : class
+    {
+        return services.ReplaceService(Mock.Of<TService>(Strict));
+    }
+    
```

Comment:
Hey, there'd not be a need for a new overload as you could just use the existing `services.ReplaceService(theMockService)` method.

Reply:
Not sure this is true. For example:

```cs
ReplaceService<ITableBuilderService>(new Mock<ITableBuilderService>()))
```

This will have a compiler error as the mock is not the same service type, and it doesn't get implicitly converted into the right type either.

In fact, if we did just this:

```cs
ReplaceService(new Mock<ITableBuilderService>()))
```

It would try and replace a service with type `Mock<ITableBuilderService>` in the container, which would just throw an error as this service won't exist.

However, if we had an overload like this:

```cs
public static IServiceCollection ReplaceService<TService>(
    this IServiceCollection services,
    Mock<TService> replacement)
    where TService : class
{
    return services.ReplaceService<TService>(replacement.Object);
}
```

It wouldn't be possible to have the above issue as the `Mock<TService>` is recognised and then unboxed beforehand. If anything, we should do this just to be safe.

Note that I've also changed my suggestion to an overload of `ReplaceService` rather than `MockService`.

Reply:
Ahh yeah of course, I was thinking of Mock.Of() as opposed to new Mock().  Cool, I've added the overload, ta.